src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_seckey.c

```diff
@@ -4,26 +4,28 @@
 
 #include "pal_seckey.h"
 
-extern "C" int32_t AppleCryptoNative_SecKeyExport(
+const static SecItemImportExportKeyParameters EmptySecItemImportExportKeyParameters;
```

Comment:
Is this really the best way to do initialize the locals later?  It can't be done at the declaration site with `= {0}`?

Reply:
If it is an array, then `{0}` initialization work in C and C++. In C++, `{}` also does the same thing.

In case of struct, `{0}` works only with C++. In C, we need to either:

* define EmptyXX, like I have done (compile time constant), all file-level static are initialized to 0
* define `(MY_TYPE){0,0,....,0}`, enter 0 number of times we have fields in struct and type cast
* memset, which is runtime memset (expensive than const)

Reply:
> Is this really the best way

Yes, I believe so.

> with `= {0}`

`{0}` is invalid for struct having more than one field.

If it is an array, then `{0}` initialization works in C and C++. In C++, `{}` also does the same thing.

In case of struct, `{}` and `{0}` work only with C++. In C, we need to either:

* define compile-time const, like I have used EmptyXX, all file-level const static are initialized to 0 (ISO C standard)
* define `(MY_TYPE){0,0,....,0}`, enter 0 number of times we have fields in struct and type cast. Hard to count and remember how many fields are there, especially for structs that we don't control and that are defined in system headers.
* Use designators to initialize struct: `(MY_TYPE){.field1: 0, field2: 0, .... fieldN: 0}`. Hard to remember each field's name in struct
* memset, which is a runtime call (expensive than const i have used)

Reply:
> memset, which is a runtime call (expensive than const i have used)

Really? I would expect most modem compilers to optimize it equivalently. Can you share a benchmark that shows it's measurably slower? It's also what we use elsewhere, and I find it makes it much clearer what's happening. With your code, I had to search for the symbol to find it and see that it was indeed left with its default init, was a const so it wasn't modified, etc.

Reply:
Ok, `memset` indeed get optimized and finally produces even less garbage when compared to EmptyXX.

I have tested by producing assembly code with:

```sh
clang -S -x c -masm=intel -o- - << EOF
#include <string.h>
#include <stdio.h>
typedef struct { int a; char b; } MY_STRUCT;
int main(void)
{
  MY_STRUCT x;
  memset(&x,0,sizeof(MY_STRUCT));
  printf("%i\n", x.a);
  return 0;
}
EOF
```

vs.

```sh
clang -S -x c -masm=intel -o- - << EOF
#include <string.h>
#include <stdio.h>
typedef struct { int a; char b; } MY_STRUCT;
const static MY_STRUCT EmptyMyStruct;
int main(void)
{
  MY_STRUCT x = EmptyMyStruct;
  printf("%i\n", x.a);
  return 0;
}
EOF
```

diff:

```diff
.Lfunc_end0:
        .size   main, .Lfunc_end0-main
        .cfi_endproc
                                        # -- End function
+         .type   EmptyMyStruct,@object   # @EmptyMyStruct
        .section        .rodata,"a",@progbits
        .p2align        2
+ EmptyMyStruct:
+         .zero   8
+         .size   EmptyMyStruct, 8
+ 
+         .type   .L.str,@object          # @.str
+         .section        .rodata.str1.1,"aMS",@progbits,1
```